src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt

```diff
@@ -0,0 +1,243 @@
+                } else {
+                    processSingleResponse(firstPart, responseStream, endStream, whatCount, count)
+                }
+            }
+        }, false,)
+            .single()
+            .onErrorResume {
+                Mono.just(
+                    SingleResponse(
+                        null,
+                        JsonRpcError(RpcResponseError.CODE_UPSTREAM_INVALID_RESPONSE, it.message ?: "Internal error"),
+                    ),
+                )
+            }
+    }
+
+    private fun processSingleResponse(
+        response: SingleResponse,
+        responseStream: Flux<ByteArray>,
+        endStream: AtomicBoolean,
+        whatCount: AtomicReference<Count>,
+        count: AtomicInteger,
+    ): Mono<out Response> {
+        if (response.noResponse()) {
+            throw IllegalStateException("Invalid JSON structure")
+        } else {
+            if (response.hasError()) {
+                return Mono.just(response)
+            } else {
+                return if (endStream.get()) {
+                    Mono.just(SingleResponse(response.result, null))
+                } else {
+                    Mono.just(
+                        StreamResponse(
+                            streamParts(
+                                response.result!!,
+                                responseStream,
+                                endStream,
+                                whatCount,
+                                count,
+                            ),
+                        ),
+                    )
+                }
+            }
+        }
+    }
+
+    private fun aggregateResponse(response: Flux<ByteArray>, statusCode: Int): Mono<AggregateResponse> {
+        return ByteBufFlux.fromInbound(response).aggregate().asByteArray()
+            .map { AggregateResponse(it, statusCode) }
+    }
+
+    private fun streamParts(
+        firstBytes: ByteArray,
+        responseStream: Flux<ByteArray>,
+        endStream: AtomicBoolean,
+        whatCount: AtomicReference<Count>,
+        count: AtomicInteger,
+    ): Flux<Chunk> {
+        return Flux.concat(
+            Mono.just(Chunk(firstBytes, false)),
+            responseStream.skip(1)
+                .filter { !endStream.get() }
+                .map { bytes ->
+                    val whatCountValue = whatCount.get()
+                    for (i in bytes.indices) {
+                        if (whatCountValue is Count.CountObjectBrackets) {
+                            countBrackets(bytes[i], count, OBJECT_OPEN_BRACKET, OBJECT_CLOSE_BRACKET)
+                        } else if (whatCountValue is Count.CountArrayBrackets) {
+                            countBrackets(bytes[i], count, ARRAY_OPEN_BRACKET, ARRAY_CLOSE_BRACKET)
+                        } else if (whatCountValue is Count.CountQuotes) {
+                            if (bytes[i] == QUOTE) {
+                                count.decrementAndGet()
```

Comment:
So, technically you can meet something like \" inside string value, so looks like this parsing could be broken

Reply:
Fixed